src/resolver.rs

```diff
@@ -1335,602 +1174,410 @@ pub fn resolve<'a>(
-    //   shouldn't logically affect any other node's results. i.e. UsesX being safe-to-run
-    //   does not depend on X's tests being safe-to-run. If we were to desugar X into two nodes
-    //   as Cargo does, the "dev" node would always be a root, so we should treat it as such
-    //   during this second pass.
-    //
-    // * We don't actually *care* if a node *has* tests/benches. All we care about is if
-    //   the node has dev-dependencies, because that's the only thing that could change the
-    //   results of the previous pass. (TODO: is this actually true? Root nodes get special
-    //   default policies, so maybe all nodes that are testable should get rescanned? But
-    //   also tests are expected to have weaker requirements... hmm...)
-    //
-    // * We will only ever test/bench a workspace member, and workspace members are always
-    //   first-party packages. This means we can avoid thinking about all the complicated
-    //   graph search stuff and just need to do simple one-shot analysis of our deps and
-    //   check against root policies.
-    for &pkgidx in &graph.topo_index {
+    // Walk the topo graph in reverse, so that we visit each package before any
+    // dependencies.
+    for &pkgidx in graph.topo_index.iter().rev() {
         let package = &graph.nodes[pkgidx];
-        trace!("resolving dev {}:{}", package.name, package.version,);
-        if package.is_workspace_member {
-            resolve_dev(
-                store,
-                graph,
-                criteria_mapper,
-                &mut results,
-                &mut violations,
-                &mut root_failures,
-                &mut root_caveats,
-                pkgidx,
-            );
-        } else {
-            assert!(
-                package.dev_deps.is_empty(),
-                "{}:{} isn't a workspace member but has dev-deps!",
-                package.name,
-                package.version
+        let policy = policy.get(package.name);
+
+        if let Some(c) = policy.and_then(|p| p.criteria.as_ref()) {
+            // If we specify a policy on ourselves, override any requirements we've
+            // had placed on us by reverse-dependencies.
+            requirements[pkgidx] = criteria_mapper.criteria_from_list(c);
+        } else if package.is_root {
+            // If this is a root crate, it will require at least
+            // `DEFAULT_POLICY_CRITERIA` by default, unless overridden.
+            requirements[pkgidx].unioned_with(
+                &criteria_mapper.criteria_from_list([format::DEFAULT_POLICY_CRITERIA]),
             );
         }
-    }
+        let normal_criteria = requirements[pkgidx].clone();
 
-    // If there were violations, report that
-    if !violations.is_empty() {
-        return (
-            results,
-            Conclusion::FailForViolationConflict(FailForViolationConflict { violations }),
-        );
+        // For each dependency,
```

Comment:
I expect there should be more to this comment. Maybe
```suggestion
        // For each dependency, elaborate the dependency criteria from the configured policy and add it to the dependency requirements.
```
Though the above section which does it first for the `dev_deps` doesn't have any such comment :shrug: I would be equally satisfied with this lingering half-comment being removed.

Reply:
Lol, that sure looks like a partial comment I got distracted while writing :p - thanks for the catch.